src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareDuplicates.java

```diff
@@ -0,0 +1,203 @@
+                }
+            }
+        }
+    }
+}
+
+
+class Utils implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    enum MatchType {
+        EQUAL,          // The same number of reads per key, same number of duplicates, and same representative read.
+        SIZE_UNEQUAL,   // Difference number of reads at this key (this means the BAMs are different).
+        READ_MISMATCH,  // The reads at this key are different for the two BAMs (which means the BAMs are different).
+        DIFF_NUM_DUPES, // The reads are the same from each BAM, but there are different number of duplicates marked.
+        DIFFERENT_REPRESENTATIVE_READ,  // The reads are the same and the same number of duplicates, but with a
+                                        // different representative read (usually this is from a tie-breaker).
+    }
+
+    static boolean isNonPrimary(GATKRead read) {
+        return read.isSecondaryAlignment() || read.isSupplementaryAlignment() || read.isUnmapped();
+    }
+
+    // As taken from part of MarkDuplicatesSparkUtils.subkeyForFragment
+    static String groupKey(GATKRead r) {
+        return r.getContig() + "," + ReadUtils.getStrandedUnclippedStart(r) + "," + ReadUtils.readHasMappedMate(r);
+    }
+
+    /**
+     * getDupes returns the metadata about how well the two sets of reads match.
+     * @param f reads from the first bam
+     * @param s reads from the second bam
+     * @param header header (should be the same for both)
+     * @return the type of the match, EQUAL, DIFFERENT_REPRESENTATIVE_READ, etc.
+     */
+    static MatchType getDupes(Iterable<GATKRead> f, Iterable<GATKRead> s, SAMFileHeader header) {
+        List<GATKRead> first = Lists.newArrayList(f);
+        List<GATKRead> second = Lists.newArrayList(s);
+        if (first.size() != second.size()) {
+            return MatchType.SIZE_UNEQUAL;
+        }
+        int size = first.size();
+        if (size > 1) {
+            first.sort(new ReadCoordinateComparator(header));
+            second.sort(new ReadCoordinateComparator(header));
+        }
+
+        HashSet<GATKRead> firstDupes = Sets.newHashSet();
+        HashSet<GATKRead> secondDupes = Sets.newHashSet();
+        for (int i = 0; i < size; ++i) {
+            GATKRead firstRead = first.get(i);
+            GATKRead secondRead = second.get(i);
+            if (!firstRead.getName().equals(secondRead.getName())) {
+                return MatchType.READ_MISMATCH;
+            }
+            if (firstRead.isDuplicate()) {
+                firstDupes.add(firstRead);
+            }
+            if (secondRead.isDuplicate()) {
+                secondDupes.add(secondRead);
+            }
+        }
+        if (firstDupes.size() != secondDupes.size()) {
+            return MatchType.DIFF_NUM_DUPES;
+        }
+        if (firstDupes.isEmpty()) {
+            return  MatchType.EQUAL;
+        }
+        if (!firstDupes.containsAll(secondDupes)) {
```

Comment:
this would be much simpler to read with just `!firstDupes.equals(secondDupes)` which amounts to the same thing but it's easier to reads because name `equals` implies symmetry and `containsAll` does not.